From: Dimitri Sivanich <sivanich@sgi.com>
Subject: mm: Allow non-rt tasks to sleep on queue for lock page

Patch-mainline: no
References: bnc#699916

This change avoids RT priority inversion issues associated with
locked pages.

Signed-off-by: Dimitri Sivanich <sivanich@sgi.com>
Signed-off-by: Mel Gorman <mgorman@suse.de>

---
 mm/filemap.c |   12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

Index: linux/mm/filemap.c
===================================================================
--- linux.orig/mm/filemap.c
+++ linux/mm/filemap.c
@@ -713,11 +713,13 @@ void __lock_page(struct page *page)
 	DEFINE_WAIT_BIT(wait, &page->flags, PG_locked);
 
 	do {
-		while (PageUptodate(page) && !PageWriteback(page) &&
-				!need_resched()) {
-			cpu_relax();
-			if (!PageLocked(page) && trylock_page(page))
-				goto done;
+		if (!rt_task(current)) {
+			while (PageUptodate(page) && !PageWriteback(page) &&
+					!need_resched()) {
+				cpu_relax();
+				if (!PageLocked(page) && trylock_page(page))
+					goto done;
+			}
 		}
 
 		prepare_to_wait(wq, &wait.wait, TASK_UNINTERRUPTIBLE);
